proposals/pattern-variables.md

```diff
@@ -0,0 +1,139 @@
+# Variable declarations under pattern combinators
+
+## Summary  
+
+## Motivation
+
+## Specification
+
+### Scope
+
+#### Pattern variable redeclaration
+
+Introducing a new concept similar to *name hiding*, but these names can possibly reference either of variables based on the result of the pattern-matching at runtime.
+
+For a *disjunctive_pattern* of the form `left_pattern or right_pattern`:
+- Pattern variables declared in the *left_pattern* can be redeclared in the *right_pattern*; permitting:
+
+  ```cs
+  e is (0, var x) or (var x, 0)
+  ```
+
+For a *logical_or_expression* of the form `left_expr || right_expr`:
+- Pattern variables declared in *left_expr* can be redeclared in the *right_expr*; permitting:
+
+  ```cs
+  e is (0, var x) || e is (var x, 0)
+  ```
+  Note: This is the expression variation of the previous case.
+
+For a *negated_pattern* of the form `not pattern_operand`:
+- Pattern variables declared in the *pattern_operand* can be redeclared anywhere in the containing expression; permitting:
+
+	```cs
+	e is not (0, var x) and not (var x, 0)
+	e is not (0, var x) && e is not (var x, 0)
+	```
+   Note: This is the DeMorgan's transformation of the previous case.
+
+For a *logical_not_expression* of the form `!expr_operand`:
+- Pattern variables declared in the *expr_operand* can be redeclared anywhere in the containing expression; permitting:
+
+  ```cs
+  !(e is (0, var x)) && !(e is (var x, 0))
+  ```
+  Note: This is the expression variation of the previous case.
+
+For a *switch_case_label* of the form `case case_pattern when when_expr`:
+- Pattern variables declared in the *case_pattern* or *when_expr* can be redeclared in other case labels; permitting:
+
+	```cs
+	case (var x, 0) when e is int i:
+	case (0, var x) when a is int i:
+	```
+
+	Note: This rule is currently defined for each individual *switch_section* only.
+	
+For a *conditional_expression* of the form `expr_cond ? expr_true : expr_false`:
+- Pattern variables declared in *expr_true* can be redeclared in *expr_false*; permitting:
+
+	```cs
+	b ? x is int i : y is int i
+	```
+
+Redeclaring pattern variables is only permitted for variables of the same type.
+
+> **Open question**: How identical these types should be?
+
+Redeclaring pattern variables in any other case follows the usual scoping rules and is disallowed.
+
+### Definite assignment
+
+Note: This section is unchanged and included for the sake of completeness.
+
+For an *is_pattern_expression* of the form `e is pattern`:
+
+- The state of *v* after *is_pattern_expression* is the same as the state of *v* after *pattern*.
+
+#### General rules for pattern variables in simple patterns 
+
+Note: This section is unchanged and included for the sake of completeness.
+
+The following rules applies to any *primary_pattern* that declares a variable:
+
+- The state of *v* is definitely assigned after *primary_pattern* if the pattern is irrefutable; permitting use after:
```

Comment:
This rule should probably be moved to the previous section under *is_pattern_expression*. Because only at the top-level we can determine if a pattern is irrefutable (at least that's how it works today).

Reply:
Consider linking to those original/unmodified sections instead of replicating them here.

Reply:
There's no one place for this in previous proposals. We have three documents each of which mentioned definite assignment in prose.

In any case, since we don't want to add expressions to the mix just yet, I think we no longer need to specify definite assignment in this format.